test/units/modules/cloud/amazon/test_data_pipeline.py

```diff
@@ -0,0 +1,275 @@
+                    setup:
+                        get_object_from_s3
+                        setup_pipeline
+                    tests:
+                        test_define_pipeline
+                        test_deactivate_pipeline
+                        test_activate_pipeline
+                        test_create_pipeline
+                        test_create_pipeline_with_tags
+                        test_create_pipeline_already_exists
+                        test_delete_nonexistent_pipeline
+                        test_delete_pipeline
+                        test_build_unique_id_different
+                        test_build_unique_id_same
+                        test_build_unique_id_obj
+                        test_format_tags
+                        test_format_empty_tags
+                        test_pipeline_description
+                        test_pipeline_description_nonexistent
+                        test_pipeline_field
+                        test_check_dp_exists_true
+                        test_check_dp_exists_false
+                        test_check_dp_status
+                    cleanup:
+                        tear_down_pipeline
+'''
+
+import pytest
+import boto3
+import os
+import json
+from . placebo_fixtures import placeboify, maybe_sleep
+from ansible.modules.cloud.amazon import data_pipeline as dp
+from ansible.modules.cloud.amazon import s3
+from ansible.module_utils._text import to_text
+
+
+class FakeModule(object):
+    def __init__(self, **kwargs):
+        self.params = kwargs
+
+    def fail_json(self, *args, **kwargs):
+        self.exit_args = args
+        self.exit_kwargs = kwargs
+        raise Exception('FAIL')
+
+    def exit_json(self, *args, **kwargs):
+        self.exit_args = args
+        self.exit_kwargs = kwargs
+
+
+def get_object_from_s3():
+    if not os.getenv('PLACEBO_RECORD'):
+        objects = [{"name": "Every 1 day",
+                    "id": "DefaultSchedule",
+                    "fields": []},
+                   {"name": "Default",
+                    "id": "Default",
+                    "fields": []}]
+        return objects
+    else:
+        s3 = boto3.client('s3')
+        data = s3.get_object(Bucket="ansible-test-datapipeline", Key="pipeline-object/new.json")
+        objects = json.loads(to_text(data['Body'].read()))
+        return objects
+
+
+def setup_pipeline(placeboify, description='', objects=[], tags={}):
+    connection = placeboify.client('datapipeline')
+    params = {'name': 'ansible-test-create-pipeline',
+              'description': description,
+              'state': 'present',
+              'timeout': 300,
+              'objects': objects,
+              'tags': tags,
+              'parameters': [],
+              'values': []}
+    m = FakeModule(**params)
+    changed, result = dp.create_pipeline(connection, m)
+    return m, connection, result['data_pipeline']['pipeline_id']
+
+
+def tear_down_pipeline(connection, module):
```

Comment:
This sort of work is best done as a pytest fixture. With [fixtures](https://docs.pytest.org/en/latest/fixture.html) you could write it such that one fixture would set up & tear down the whole setup. 

```
@pytest.fixture
def pipeline_fixture(placeboify, objects, tags):
    connection = placeboify.client('datapipeline')
    [insert params stuff and FakeModule]
    _, result = db.create_pipeline(connection, m)
    yield {'module': m, 'connection': connection, 'pipeline_id': result['data_pipeline']['pipeline_id']}
    
    # this part runs after the test finishes, pass or fail
    dp.delete_pipeline({'name': 'ansible-test-create-pipeline', 'state': 'absent'})
```

And that would merge the setup/teardown functions into a fixture that you could use in tests like:

```
test_define_pipeline(pipeline_fixture, maybe_sleep):
    pipeline_fixture['module'].[call a method or what have you](pipeline_fixture['pipeline_id'])
```